### PR TITLE
[CodeCompletion] Improve accuracy of unresolved member completion

### PR DESCRIPTION
**Pull request by rintaro**
Follow up to https://github.com/apple/swift/pull/19567#issuecomment-424909508

* Handle generic base types
* Suggests `.some` and `.none` for optional types
* Don't look through too many parameter lists for function types
* Include members with "convertible" type result. Previously it was "equals"

rdar://problem/44803439

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
I'm not sure these are officially supported:
```swift
// Implicit member expression via member type construction.
class Base {
  class Derived : Base {}
}
let _: Base = .Derived()

// Implicit member expression via calling closure type static property.
struct Foo {
  static var factory: () -> Foo = { fatalError() }
}
let _: Foo = .factory()
```
I included them because the current compiler allows them.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@rintaro If they're allowed they're "supported". We don't have a formal spec :) At least the nested type case is something I've seen in real code before. Can you make sure they're tested somewhere in `test/expr/dot` as well?

File: test/IDE/complete_unresolved_members.swift [hunk 1]
```diff
@@ -188,12 +199,33 @@ class C4 {
...
+// UNRESOLVED_3_OPT-DAG: Decl[EnumElement]/ExprSpecific:     none[#Optional<Wrapped>#]; name=none
+// UNRESOLVED_3_OPT-DAG: Decl[EnumElement]/ExprSpecific:     some({#Wrapped#})[#(Wrapped) -> Optional<Wrapped>#];
+// UNRESOLVED_3_OPT-DAG: Decl[Constructor]/CurrNominal:      init({#(some): SomeEnum1#})[#Optional<SomeEnum1>#];
+// UNRESOLVED_3_OPT-DAG: Decl[Constructor]/CurrNominal:      init({#nilLiteral: ()#})[#Optional<SomeEnum1>#];
+
+// UNRESOLVED_3_OPTOPTOPT: Begin completions
+// UNRESOLVED_3_OPTOPTOPT-DAG: Decl[EnumElement]/ExprSpecific:     North[#SomeEnum1#];
+// UNRESOLVED_3_OPTOPTOPT-DAG: Decl[EnumElement]/ExprSpecific:     South[#SomeEnum1#];
+// UNRESOLVED_3_OPTOPTOPT-DAG: Decl[EnumElement]/ExprSpecific:     none[#Optional<Wrapped>#]; name=none
+// UNRESOLVED_3_OPTOPTOPT-DAG: Decl[EnumElement]/ExprSpecific:     some({#Wrapped#})[#(Wrapped) -> Optional<Wrapped>#];
```
Comment (benlangmuir) [inline test/IDE/complete_unresolved_members.swift]: Do we know why the generic `Wrapped` isn't resolved here?

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**